### PR TITLE
Mirror matrix

### PR DESCRIPTION
**Pull request by msimberg**
For #319 there is a need for transferring matrices between host and device. Usually this needs to not only be done once, but the copies need to be synchronized. This issue is meant to discuss options.

What I would propose is a `MatrixMirror` type that handles the copying. An important requirement is that the mirror does not make any copies if the source and target devices are the same. In that case the mirror would act exactly as a reference to the original matrix. Sketch:

```
template <class T, Device Target, Device Source>
class MatrixMirror {
    // Store the source matrix as a reference. The source can still be accessed.
    // This creates a copy (if needed) on the target device.
    MatrixMirror(Matrix<T, Source> const&);
    // This copies matrix back from the target to the source.
    ~MatrixMirror();

    // Synchronize explicitly from target to source.
    void syncTargetToSource();
    // Synchronize explicitly from source to target.
    void syncSourceToTarget();

    // Access the target (copy) matrix.
    future<TileType> operator()(LocalTileIndex);
    shared_future<TileType> read(LocalTileIndex);
    // And the same for global tile indices...
};
```

In the last commit on this branch I've only implemented a `HostMatrix` (should really have mirror in the name...) which is a mirror of a source matrix on `Device::CPU` memory. I think the more general formulation from above is probably a better choice. Typedefs could be introduced for the common case of the source or target being `Device::CPU`. This would need to conform to the `MatrixLike` template template parameters that are used in a few places.

Note that an independent feature would be to implement `set` for `Device::GPU` matrices as well.  I'm doing this first for a few reasons: 1. I'm not convinced about the CUDA executor in HPX being the right choice, and 2. the `MatrixMirror` covers that case as well (although with copies), and 3. gtest checks should be done on the host, so I would anyway need this feature to transfer device matrices to the host. This feature allows doing that generically.

Unfortunately builds on top of #319 because that's where I get to test this, but the feature itself is in principle independent of it.

### CODE REVIEW

**Comment by rasolca** [comment]
A few remarks that come to my mind:
- I think you want to reuse the same mirror matrix in multiple part of an algorithm, therefore I think you should be able to sync just a submatrix (e.g. specifying the index of the top left tile and the number (2D size) of tiles)

- My initial idea was to just use a CPU and a GPU Matrix and use `copy` method of matrices. In this way there is a full control of the copies using either matrix copy or tile by tile copy ( `copy` can be easily extended to deal with submatrices and triangular matrices). Did you consider this approach?

**Comment by msimberg** [comment]
> A few remarks that come to my mind:
> 
>     * I think you want to reuse the same mirror matrix in multiple part of an algorithm, therefore I think you should be able to sync just a submatrix (e.g. specifying the index of the top left tile and the number (2D size) of tiles)
Good idea. I admit my initial intention with this was mostly for higher-level ("user code") rather than within an algorithm. 
> 
>     * My initial idea was to just use a CPU and a GPU Matrix and use `copy` method of matrices. In this way there is a full control of the copies using either matrix copy or tile by tile copy ( `copy` can be easily extended to deal with submatrices and triangular matrices). Did you consider this approach?
This is essentially what the triangular solver miniapp and the tile trsm tests do right now. To avoid duplication I'd like to combine them into one generic version. If there is a generic version I'd like to avoid having two copies of a matrix on the CPU when in CPU-only mode or needing ifdefs in the code to disable the copy when not needed. To do that some sort of higher-level abstraction is needed that knows about the connection between the CPU and (potentially) GPU matrix. I hope I'm not misunderstanding what you're suggesting here.

**Comment by rasolca** [comment]
Sorry, I though it was a functionality to simplify the implementation of algorithms. If it is meant for user code, it LGTM.

**Comment by msimberg** [comment]
I've done some more changes. The main `MatrixMirror` type is now working quite nicely. There are a few unit tests.

What I'm still unsure about is how and if to template everything that takes a `Matrix` and could potentially take a `MatrixMirror` as well. I currently have two options:
1. Actually allow passing a `MatrixMirror` to everything that takes a `Matrix`. Because of the additional `Device` template parameter in `MatrixMirror` functions need to be in the simplest case be templated like:
```
template <typename MatrixType>
void f(MatrixType& mat);
```
but in the worst case like:
```
template <template <class, Device, Device...> class MatrixLike, class T, Device D, Device... Ds>
void f(MatrixLike<T, D, Ds...>& mat);
```
2. Just add a `get()` method to `MatrixMirror` which returns the actual `Matrix`.

From an interface perspective I would really like to do 1. but I find it very hard to justify as it's a pain to read. Especially for functions that take multiple matrices there should in principle be a separate `MatrixLike` template template parameter for each matrix if each input matrix can be a different `MatrixLike`.

What do you think? Is 1. fine? Is having to call `get()` annoying or nicely explicit? Unfortunately implicit conversions are not an option since they're not considered for templated functions. Can you think of other options? For some functions explicit `MatrixMirror` overloads are also an option, but I don't like the duplication it causes.

Note that my latest push contains a mix of these ideas, and needs cleaning up.

**Comment by rasolca** [comment]
I think that approach 1.a is more elegant to use for the public API, but I would use approach 2 for the internal API. 

In fact, if approach 1.a is applied to the internal API as well, it can cause some problems:
- ETI: the same function will be added to the library twice (once for Matrix and once for MirrorMatrix)
- can mix datatype and devices, and therefore requires static_asserts to check the consistency.

**Comment by msimberg** [comment]
Just be sure we're talking about the same thing, you mean by 1.a this:
```
template <typename MatrixType>
void f(MatrixType& mat);
```
and implicitly 1.b. is this:
```
template <template <class, Device, Device...> class MatrixLike, class T, Device D, Device... Ds>
void f(MatrixLike<T, D, Ds...>& mat);
```
?

What you're suggesting is to turn this:
```
template <Backend backend, Device device, class T>
void triangular(blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag, T alpha,
                Matrix<const T, device>& mat_a, Matrix<T, device>& mat_b);
```
into this:
```
template <Backend backend, class MatrixType1, class MatrixType2>
void triangular(blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag, T alpha,
                MatrixType1& mat_a, MatrixType2& mat_b) {
// when dispatching
internal::Triangular<backend, device, T>::call_LLN(diag, alpha, getMatrix(mat_a), getMatrix(mat_b));
}
```
?

I think it'd preferable to have as precise types as possible at the API level (e.g. `mat_a` is read-only and `mat_b` is read-write), but the signature would of course be much more complicated. This is why I'm (very slightly) in favour of only allowing `Matrix` in these interfaces. The whole discussion becomes different if we foresee other types than just `MatrixMirror` being possible to pass to the algorithms (`MatrixView`? I get the impression that's more for internal use).

**Comment by rasolca** [comment]
> Just be sure we're talking about the same thing, you mean by 1.a this:
> 
> ```
> template <typename MatrixType>
> void f(MatrixType& mat);
> ```
> 
> and implicitly 1.b. is this:
> 
> ```
> template <template <class, Device, Device...> class MatrixLike, class T, Device D, Device... Ds>
> void f(MatrixLike<T, D, Ds...>& mat);
> ```
> 
> ?

Yes, exactly.

> What you're suggesting is to turn this:
> 
> ```
> template <Backend backend, Device device, class T>
> void triangular(blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag, T alpha,
>                 Matrix<const T, device>& mat_a, Matrix<T, device>& mat_b);
> ```
> 
> into this:
> 
> ```
> template <Backend backend, class MatrixType1, class MatrixType2>
> void triangular(blas::Side side, blas::Uplo uplo, blas::Op op, blas::Diag diag, T alpha,
>                 MatrixType1& mat_a, MatrixType2& mat_b) {
> // when dispatching
> internal::Triangular<backend, device, T>::call_LLN(diag, alpha, getMatrix(mat_a), getMatrix(mat_b));
> }
> ```
> 
> ?

This is one option. What I wanted to say is that allowing multiple types in the internal API is not ideal.

> I think it'd preferable to have as precise types as possible at the API level (e.g. `mat_a` is read-only and `mat_b` is read-write), but the signature would of course be much more complicated. This is why I'm (very slightly) in favour of only allowing `Matrix` in these interfaces. The whole discussion becomes different if we foresee other types than just `MatrixMirror` being possible to pass to the algorithms (`MatrixView`? I get the impression that's more for internal use).

I agree, if you want to keep precise types I suggest to use approach 2 for the public API, otherwise the signature might become cumbersome.
Yes, MatrixView is meant to be used, when needed, for the internal API.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Review by rasolca — COMMENTED** [review]
A small note about the triangular solver test:
Don't know why, but the test is not looping on all the communicators.
Please fix the following line:
https://github.com/eth-cscs/DLA-Future/blob/c34e03368d51dbb8f3e8048fa8222446df102e8f/test/unit/solver/test_triangular.cpp#L175

File: test/unit/solver/test_triangular.cpp [hunk 1]
```diff
@@ -169,7 +185,12 @@ TYPED_TEST(TriangularSolverDistributedTest, Correctness) {
               std::tie(m, n, mb, nb) = sz;
               TypeParam alpha = TypeUtilities<TypeParam>::element(-1.2, .7);
 
-              testTriangularSolver<TypeParam>(comm_grid, side, uplo, op, diag, alpha, m, n, mb, nb);
+              testTriangularSolver<TypeParam, Backend::MC, Device::CPU>(comm_grid, side, uplo, op, diag,
+                                                                        alpha, m, n, mb, nb);
+#ifdef DLAF_WITH_CUDA
+              testTriangularSolver<TypeParam, Backend::GPU, Device::GPU>(comm_grid, side, uplo, op, diag,
+                                                                         alpha, m, n, mb, nb);
+#endif
```
[the remark this answers is not in the document]
Comment (msimberg) [reply test/unit/solver/test_triangular.cpp]: Done.

**Comment by msimberg** [comment]
> A small note about the triangular solver test:
> Don't know why, but the test is not looping on all the communicators.
> Please fix the following line:
> 
> https://github.com/eth-cscs/DLA-Future/blob/c34e03368d51dbb8f3e8048fa8222446df102e8f/test/unit/solver/test_triangular.cpp#L175

I can't explain that one, but very good catch. Changed that.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by rasolca** [comment]
Please note, that if you change the name of the test suite (the first argument to TEST macros) you should also modify https://github.com/eth-cscs/DLA-Future/blob/716955383099b44e5d3b7a89f86575299d0bd492/test/unit/solver/test_triangular.cpp#L34-L45

**Comment by msimberg** [comment]
Yeah, I just realized that and pushed another change. I made them
```
TYPED_TEST(TriangularSolverTest, CorrectnessLocalMC)
TYPED_TEST(TriangularSolverTest, CorrectnessDistributedMC)
TYPED_TEST(TriangularSolverTest, CorrectnessLocalGPU)
TYPED_TEST(TriangularSolverTest, CorrectnessDistributedGPU)
```
Do you have a preference on which way to do it?

**Comment by rasolca** [comment]
I slightly prefer to have different suites for MC and GPU as the output group them:
```
test 37
      Start 37: test_triangular
37: Test command: /DLA-Future-build/test/unit/solver/test_triangular "--hpx:use-process-mask"
37: Test timeout computed to be: 10000000
37: Running main() from gtest_mpihpx_main.cpp
37: [==========] Running 16 tests from 8 test suites.
37: [----------] Global test environment set-up.
37: [----------] 2 tests from TriangularSolverLocalTest/0, where TypeParam = float
37: [ RUN      ] TriangularSolverLocalTest/0.CorrectnessMC
37: [       OK ] TriangularSolverLocalTest/0.CorrectnessMC (2673 ms)
37: [ RUN      ] TriangularSolverLocalTest/0.CorrectnessGPU
37: [       OK ] TriangularSolverLocalTest/0.CorrectnessGPU (2452 ms)
37: [----------] 2 tests from TriangularSolverLocalTest/0 (5186 ms total)
37: 
37: [----------] 2 tests from TriangularSolverLocalTest/1, where TypeParam = double

...

37: [----------] 2 tests from TriangularSolverDistributedTest/0, where TypeParam = float
37: [ RUN      ] TriangularSolverDistributedTest/0.CorrectnessMC
37: [       OK ] TriangularSolverDistributedTest/0.CorrectnessMC (109 ms)
37: [ RUN      ] TriangularSolverDistributedTest/0.CorrectnessGPU
37: [       OK ] TriangularSolverDistributedTest/0.CorrectnessGPU (313 ms)
37: [----------] 2 tests from TriangularSolverDistributedTest/0 (615 ms total)
37: 
37: [----------] 2 tests from TriangularSolverDistributedTest/1, where TypeParam = double
...
```
On the other hand I would merge local/distributed in one suite.

**Comment by msimberg** [comment]
> I slightly prefer to have different suites for MC and GPU as the output group them

I see now that this is exactly what you suggested in the first place. I'll go with that then.

**Comment by msimberg** [comment]
bors try

**Review by rasolca — APPROVED** [review]

**Pull request merged**